compiler/rustc_mir/src/transform/const_goto.rs

```diff
@@ -0,0 +1,135 @@
+//! }
+//! ```
+//! into
+//! ```rust
+//! bb2: {
+//!     _2 = const true;
+//!     goto -> bb5;
+//! }
+//! ```
+
+use crate::transform::MirPass;
+use rustc_middle::mir::*;
+use rustc_middle::ty::TyCtxt;
+use rustc_middle::{mir::visit::Visitor, ty::ParamEnv};
+
+use super::simplify::{simplify_cfg, simplify_locals};
+
+pub struct ConstGoto;
+
+impl<'tcx> MirPass<'tcx> for ConstGoto {
+    fn run_pass(&self, tcx: TyCtxt<'tcx>, body: &mut Body<'tcx>) {
+        trace!("Running ConstGoto on {:?}", body.source);
+        let param_env = tcx.param_env_reveal_all_normalized(body.source.def_id());
+        let mut opt_finder =
+            ConstGotoOptimizationFinder { tcx, body, optimizations: vec![], param_env };
+        opt_finder.visit_body(body);
+        let should_simplify = !opt_finder.optimizations.is_empty();
+        for opt in opt_finder.optimizations {
+            let bbd_with_goto = &mut body.basic_blocks_mut()[opt.bb_with_goto];
+            bbd_with_goto.statements.extend(opt.statements_to_append_bb);
+            let terminator = bbd_with_goto.terminator_mut();
+            let new_goto = TerminatorKind::Goto { target: opt.target_to_use_in_goto };
+            debug!("SUCCESS: replacing `{:?}` with `{:?}`", terminator.kind, new_goto);
+            terminator.kind = new_goto;
+        }
+
+        // if we applied optimizations, we potentially have some cfg to cleanup to
+        // make it easier for further passes
+        if should_simplify {
+            simplify_cfg(body);
+            simplify_locals(body, tcx);
+        }
+    }
+}
+
+impl<'a, 'tcx> Visitor<'tcx> for ConstGotoOptimizationFinder<'a, 'tcx> {
+    fn visit_terminator(&mut self, terminator: &Terminator<'tcx>, location: Location) {
+        let _: Option<_> = try {
+            let target = terminator.kind.as_goto()?;
+            // We only apply this optimization if the last statement is a const assignment
+            let last_statement = self.body.basic_blocks()[location.block].statements.last()?;
+
+            if let (place, Rvalue::Use(Operand::Constant(_const))) =
+                last_statement.kind.as_assign()?
+            {
+                // We found a constant being assigned to `place`.
+                // Now check that the target of this Goto switches on this place.
+                let target_bb = &self.body.basic_blocks()[target];
+
+                let target_bb_terminator = target_bb.terminator();
+                let (discr, switch_ty, targets) = target_bb_terminator.kind.as_switch()?;
+                if discr.place() == Some(*place) {
+                    // We now know that the Switch matches on the const place.
+
+                    // We only allow the optimization if the place being switched on is
+                    // not used in the statements in the target basic block.
+                    if is_local_assigned_in_statements(place.local, &target_bb.statements) {
+                        None?
+                    }
```

Comment:
Compared to the earlier version the transform now supports switch blocks with statements. As a result it has to establish that the definition `place = const` reaches the switch, whereas previously it was trivially true.

Based on past experiences, I feel that doing this for an arbitrary place without a dedicated analysis will be challenging. If place is indirect it could be invalidated by other assignments, it can also contain indexing projections referencing other locals that could be modified, etc. 

I would probably start with locals without any projections, and in addition to the existing check, also verify that either they never have their address taken or there are no indirect assignment in the switch block.

Reply:
Yeah, you are right. I'll revert the newest changes so that we only support statement-less targets. Then follow ups can improve the situation